### ISSUE TITLE
[Cypress version 4.3.0]TypeError: Cannot read property 'key' of undefined 

### ISSUE THREAD

**Issue by natkrish** [cypress-io/cypress#6890]
@flotwig 
### Current behavior:
So when I load our live site using cy.visit('https://onlinedoctor.lloydspharmacy.com') using cypress version 4.3.0 - I get the below error. 
![image](https://user-images.githubusercontent.com/12541984/78015731-ba118400-7341-11ea-9b27-31309a82da9e.png)
![image](https://user-images.githubusercontent.com/12541984/78015791-cf86ae00-7341-11ea-983d-130c1523958d.png)

### Desired behavior:

I expect to be able to load our public site without any issues. 

### Test code to reproduce

I think to repproduce this issue just simply access our live site using cy.visit('https://onlinedoctor.lloydspharmacy.com');
Please try and replicate with the version set up 

### Versions

Cypress version - 4.3.0
Chrome - 80
Windows 10 - 64 bit
Node 12.16.1

I was expecting for it to work after the bug fix for this ticket - https://github.com/cypress-io/cypress/issues/5602

![image](https://user-images.githubusercontent.com/12541984/78016259-779c7700-7342-11ea-8fb1-65ddbcf1d1ff.png)

**Comment by CypressCecelia** [comment]
I was not able to replicate. I am able to visit the URL in my test runner on 4.3.0. Are you possible logged in on your version of Chrome? The error says the URL it is trying to visit is followed by /logout, so I'm thinking you are being redirected in some way. You may have better luck setting the `baseURL` in your `cypress.config` file.

<img width="1216" alt="Screen Shot 2020-03-31 at 4 51 37 PM" src="https://user-images.githubusercontent.com/60482585/78074033-f4921580-736f-11ea-85dc-25f77de7e6ef.png">

**Comment by abasau-incomm** [comment]
@CypressCecelia I am also having the same issue. It appears that both my web site and https://onlinedoctor.lloydspharmacy.com/ are protected by Incapsula (https://www.imperva.com/imperva-incapsula-faq/). Incapsula generates cookies that Cypress fails to parse. Here is code from `%AppData%\Local\Cypress\Cache\4.3.0\Cypress\resources\app\packages\server\lib\request.js` which is failing:
```
      setCookiesOnBrowser: function(res, resUrl, automationFn) {
        ...
        return Promise.map(cookies, function(cyCookie) {
          var cookie, expiry;
          cookie = tough.Cookie.parse(cyCookie, {
            loose: true
          });
          debug('parsing cookie %o', {
            cyCookie: cyCookie,
            toughCookie: cookie
          });
          cookie.name = cookie.key; \\ <<<< This fails because cookie is undefined.
```
Here is Cypress log (I changed some random values for security reasons):
```
  cypress:server:request received status code & headers on request { requestId: 'request58', statusCode: 200, headers: { 'content-type': 'text/html; charset=utf-8', 'set-cookie': [ ..., '___utmvaFvuoaRv=ScK\u0001oacs; path=/; Max-Age=900', '___utmvbFvuoaRv=TZI    XasJslrv: TtI; path=/; Max-Age=900' ] } } +138ms
  cypress:server:request setting cookies on browser { url: '<my url>', defaultDomain: '<my domain>', cookies: [ ..., '___utmvaFvuoaRv=ScK\u0001oacs; path=/; Max-Age=900', '___utmvbFvuoaRv=TZI    XasJslrv: TtI; path=/; Max-Age=900' ] } +3ms
  cypress:server:request parsing cookie { cyCookie: '___utmvaFvuoaRv=ScK\u0001oacs; path=/; Max-Age=900', toughCookie: undefined } +0ms
```
As you can see `tough.Cookie.parse()` didn't parse the cookie and `toughCookie: undefined`. I suspect it doesn't like `\u0001` in cookie value.

**Comment by jennifer-shehane** [comment]
@abasau-incomm @natkrish - Can you downgrade to Cypress 4.2.0 and verify if this fixes the issue?

@natkrish (For my future notes - I had to VPN in to US because their website is blocked in my country).

I'm not able to reproduce the error with the given code - please provide all `cypress.json`, any plugins/support files, etc that are necessary to encounter the error when running the tests - likely there is some cookies being set when you login - then logout that will reproduce the error. Please provide a fully reproducible example.

@abasau-incomm I don't see an exact change to these lines of code, although I do see that tough-cookie was updated in 4.3.0 here: https://github.com/cypress-io/cypress/pull/6778/files#diff-8c42f42f65d0890e0bba62a07ab08042R108 Please also if you can provide a reproducible example - this would be helpful.

**Comment by natkrish** [comment]
@jennifer-shehane @abasau-incomm @CypressCecelia As per @abasau-incomm  said my site is protected by Incapsula. I am not sure why this happens and every time I come across this type of issue on production - I then had to downgrade to 3.4.1 which is the only version that doesn't throw this error. 
However on the other hand after @CypressCecelia  said she was able to access without any issues, I created a new project from scratch and I was able to access my production site with out any issues. See this 
![image](https://user-images.githubusercontent.com/12541984/78128994-c1eb2a00-740e-11ea-9d3c-3513591cd004.png)
But as @jennifer-shehane it will be great if you can infer any thing from these files that might hinder the launch of my site - it will be good to get down to the bottom of the problem. 
### cypress/plugins/index.js
// ***********************************************************
// This example plugins/index.js can be used to load plugins
//
// You can change the location of this file or turn off loading
// the plugins file with the 'pluginsFile' configuration option.
//
// You can read more here:
// https://on.cypress.io/plugins-guide
// ***********************************************************

// This function is called when a project is opened or re-opened (e.g. due to
// the project's config changing)
const cucumber = require('cypress-cucumber-preprocessor').default;

function getCliArgs(args) {
    args = args.filter((arg) => {
        return arg !== "--use-fake-ui-for-media-stream" && arg !== "--use-fake-device-for-media-stream"
    });
    args.push('--allow-file-access-from-files');
    return args
}

module.exports = (on, config) => {
    /** the rest of your plugins... **/
    require('cypress-log-to-output').install(on);
    on('file:preprocessor', cucumber());
    // `on` is used to hook into various events Cypress emits
    // `config` is the resolved Cypress config
    on('before:browser:launch', (browser, args) => {
        if (browser.family === 'chrome') {
            // `args` is a list of all the arguments that will be passed to Chrome when it launchers
            return getCliArgs(args)
        } else if (browser.family === 'electron') {
            // `args` is a `BrowserWindow` options object
            // https://electronjs.org/docs/api/browser-window#new-browserwindowoptions
            if (!args.webPreferences) {
                args.webPreferences = {}
            }
            if (!args.webPreferences.additionalArguments) {
                args.webPreferences.additionalArguments = []
            }
            args.webPreferences.additionalArguments = getCliArgs(args.webPreferences.additionalArguments);
            return args
        }
    })
};
### cypress.json
{
  "chromeWebSecurity": false,
  "reporter": "mochawesome",
  "reporterOptions": {
    "overwrite": true,
    "html": false,
    "json": true,
    "timestamp": "mmddyyyy_HHMMss"
  },
  "projectId": "4zqjtd",
  "modifyObstructiveCode": false,
  "pageLoadTimeout": 60000,
  "baseUrl": "https://onlinedoctor.lloydspharmacy.com/",
  "nodeVersion":"system"
}
### cypress/support/commands.js
{
  "chromeWebSecurity": false,
  "reporter": "mochawesome",
  "reporterOptions": {
    "overwrite": true,
    "html": false,
    "json": true,
    "timestamp": "mmddyyyy_HHMMss"
  },
  "projectId": "4zqjtd",
  "modifyObstructiveCode": false,
  "pageLoadTimeout": 60000,
  "baseUrl": "https://onlinedoctor.lloydspharmacy.com/",
  "nodeVersion":"system"
}

### cypress/support/index.js
// ***********************************************************
// This example support/index.js is processed and
// loaded automatically before your test files.
//
// This is a great place to put global configuration and
// behavior that modifies Cypress.
//
// You can change the location of this file or turn off
// automatically serving support files with the
// 'supportFile' configuration option.
//
// You can read more here:
// https://on.cypress.io/configuration
// ***********************************************************

// Import commands.js using ES2015 syntax:
import '@bahmutov/cy-api/support'
import './commands'
require('cypress-plugin-retries');
const addContext = require('mochawesome/addContext');
Cypress.on('test:after:run', (test, runnable) => {
    if (test.state === 'failed') {
        const screenshotFileName = `${runnable.parent.title} -- ${test.title} (failed).png`;
        addContext({ test }, `assets/${Cypress.spec.name}/${screenshotFileName}`)
    }
});
Cypress.on('uncaught:exception', (err, runnable) => {
    // returning false here prevents Cypress from
    // failing the test
    return false
});

// Alternatively you can use CommonJS syntax:
// require('./commands')

**Comment by abasau-incomm** [comment]
@jennifer-shehane I checked cypress 4.2.0 and a similar issue is reproducible there (https://github.com/cypress-io/cypress/issues/5602) - Cypress fails with `Error: Parse Error: Invalid header value char` error when it encounters a cookie with Unicode character in its value.

Also I found a stable way to reproduce the issue. You will need to:
1. Install and open Fiddler.
2. Uncheck File->"Capture Traffic" (we don't need to capture all traffic).
3. Go to FiddlerScript tab and modify `OnBeforeResponse()` so it looks like this:
```
public static void OnBeforeResponse(Session oSession)
{
	if (m_Hide304s && oSession.responseCode == 304)
	{
		oSession["ui-hide"] = "true";
	}

	oSession.oResponse["Set-Cookie"] = "___utmvaFvuoaRv=TkE\u0001sCvZ; path=/; Max-Age=900";

}
```
![image](https://user-images.githubusercontent.com/54558010/78174440-1b9f2480-740e-11ea-8824-7e263887954e.png)
4. Open command line prompt.
5. Configure Cypress to use Fiddler as Proxy:
```
set HTTP_PROXY=http://localhost:8888
```
6. Open Cypress:
```
npx cypress open
```
7. Modify our Cypress test to make a GET request. E.g.:
```
cy.request({ method: 'GET', url: 'https://onlinedoctor.lloydspharmacy.com/' });
```
8. Run the test. => 
![image](https://user-images.githubusercontent.com/54558010/78174974-feb72100-740e-11ea-8765-e3343d573b69.png)

**Comment by tmonteith** [comment]
I can definitely reproduce this within our code base.

Here is the failing step on version 4.3.0:
<img width="808" alt="Screen Shot 2020-04-05 at 3 42 53 AM" src="https://user-images.githubusercontent.com/22229966/78472777-6e1f6000-76f0-11ea-9e5b-0f43974dc65d.png">


And here is the failing step on version 4.2.0
<img width="733" alt="Screen Shot 2020-04-05 at 3 40 28 AM" src="https://user-images.githubusercontent.com/22229966/78472789-78d9f500-76f0-11ea-957c-375df8d1c314.png">

The same exact step is able to pass on version 3.4.1

Here is our code:
`Cypress.Commands.add("login", (email, password) => {
	cy.request({
		url: '/account/login',
		method: 'POST',
		headers: {
			Accept: 'application/json',
			'Content-Type': 'application/json;charset=UTF-8' },
		body: {
			EmailAddress: email, 
			Password: password
		},
		followRedirect: true
	})
})`

System info:
MacOS 10.15.2
Node v12.16.0
Chrome - 80

Note: Our entire test site is behind Incapsula and NGINX but every other request/page is able to load just fine on v4.3.0 which contains fix #5988 that addressed #5602 which we originally posted as a parse error issue as well. I would not be surprised if there was some sort of correlation.

Thanks!

**Comment by flotwig** [comment]
This issue occurs when `tough-cookie` is unable to parse the cookie's value.

`tough-cookie` expects cookies to follow [RFC 6265](https://tools.ietf.org/html/rfc6265#section-4.1.1), which doesn't permit control characters like `\u0001` in cookie values.

Even if `tough-cookie` is patched to allow invalid control characters, further down the line, Chrome DevTools Protocol will fail on the `Network.setCookie` call:


```
Network.setCookie failed to set cookie: {"domain":"localhost","path":"/","secure":false,"httpOnly":false,"expires":1586192011.465,"name":"___utmvaFvuoaRv","value":"TkE\u0001sCvZ"}
```

I was curious about how real browsers react to headers with invalid control characters, so I tried visiting this controller in Firefox and Chrome:

```coffee
  app.get "/invalidControlCharCookie", (req, res) ->
    ## `http` lib throws an error if we use .setHeader to set this
    res.connection.end("""
    HTTP/1.1 200 OK
    Content-Type: text/html
    Set-Cookie: ___utmvaFvuoaRv=TkE\u0001sCvZ; path=/; Max-Age=900
    Set-Cookie: _valid=true; path=/; Max-Age=900

    foo
    """)
```

Chrome 80 complains of invalid Set-Cookie headers and only sets the second cookie: 
![image](https://user-images.githubusercontent.com/1151760/78583758-42da6500-7805-11ea-9e69-1188bf48f6ba.png)

![image](https://user-images.githubusercontent.com/1151760/78583793-4c63cd00-7805-11ea-9220-0275c2eca4dc.png)

Firefox 75 appears to receive the cookie, but fails to actually set it in storage:
![image](https://user-images.githubusercontent.com/1151760/78583914-77e6b780-7805-11ea-97a2-626165976537.png)

![image](https://user-images.githubusercontent.com/1151760/78583963-859c3d00-7805-11ea-9a08-b1eef0429610.png)

In light of this, I think the best way to resolve this issue is to make Cypress ignore invalid cookie values in requests, just like browsers do.
